### ISSUE TITLE
[BUG] Increasing memory consumption of bokeh server (part 2)

### ISSUE THREAD

**Issue by EfremBraun** [bokeh/bokeh#11477]
I'm still experiencing the exact same issue mentioned in the first post of #7468. #7604 did not seem to help in my tests. It seems that Bokeh is still not releasing memory upon destruction of a session.

Using the sliders.py source code directly from https://github.com/bokeh/bokeh/blob/branch-2.4/examples/app/sliders.py, and running the following protocol:

1. Run mprof run bokeh serve sliders.py
2. Wait ~5 seconds
3. Open up 1 browser session.
4. Wait ~5 seconds
5. Open up 9 more browsers sessions quickly.
6. Wait ~2 minutes.
7. Close all 10 browser sessions.
8. Wait ~2 minutes.
9. Kill server with Ctrl-C.

I get the following using Bokeh 2.3.3:
![memory2 3 3](https://user-images.githubusercontent.com/6618019/128221545-ae897d2c-2328-4310-b6de-ad58d190775c.png)

Based on https://discourse.bokeh.org/t/automatically-destroy-sessions-after-period-of-time-bokeh-serve/6373, I decided to re-run the test with Bokeh 1.4.0. I'm still seeing the same issue:
![memory1 4 0](https://user-images.githubusercontent.com/6618019/128221640-95eb0ce6-9aca-4682-94bf-9a7c5990dac5.png)

I'm using Python 3.8.5 on a MacBook Pro running Catalina 10.15.7. I open the server sessions in Google Chrome 92.0.4515.107.

**Comment by bryevdv** [comment]
Quickl look note, here is a partial log with memory and stats logging enabled:
<details>

```
2021-08-04 09:30:19,229 [pid 99813] 0 clients connected
2021-08-04 09:30:19,230 [pid 99813]   /sliders has 17 sessions with 17 unused
2021-08-04 09:30:19,230 [pid 99813] Memory usage: 89.00 MB (RSS), 8261.00 MB (VMS)
2021-08-04 09:30:19,250   uncollected Documents: 18
2021-08-04 09:30:19,266   uncollected Sessions: 17
2021-08-04 09:30:19,291   uncollected Models: 613
2021-08-04 09:30:19,291 Scheduling 17 sessions to discard
2021-08-04 09:30:19,291 Discarding session 'qeZaLe5NBjl3m9DBdvy3XtMmIKi7RJpWxpgX9Uuvgk0N' last in use 25768.368845000005 milliseconds ago
2021-08-04 09:30:19,291 Deleting 1 modules for <bokeh.document.document.Document object at 0x19f5d0940>
2021-08-04 09:30:19,335 Discarding session 'xXbzPSsELBixkD3lv3ngumZtp5aU424RDNE3JjuO6SYD' last in use 25513.672407 milliseconds ago
2021-08-04 09:30:19,335 Deleting 1 modules for <bokeh.document.document.Document object at 0x19f6fb460>
2021-08-04 09:30:19,372 Discarding session 'rQJkqyq2kU024LK6arStkpHvKJ7Wv1Q24QKnjDFxWZ7m' last in use 25362.477251000004 milliseconds ago
2021-08-04 09:30:19,372 Deleting 1 modules for <bokeh.document.document.Document object at 0x19f7376a0>
2021-08-04 09:30:19,407 Discarding session 'E4QW9C8TJJCpE1ml7OhlSJiRlSBtMq3Xnuzvr1JynuHh' last in use 25049.893545 milliseconds ago
2021-08-04 09:30:19,407 Deleting 1 modules for <bokeh.document.document.Document object at 0x19f758f70>
2021-08-04 09:30:19,440 Discarding session 'zSFgYrFly5IoHrxmueLBHeM1f4DER8907bNFg1iV92q7' last in use 24892.953624 milliseconds ago
2021-08-04 09:30:19,440 Deleting 1 modules for <bokeh.document.document.Document object at 0x19f780880>
2021-08-04 09:30:19,474 Discarding session 'stWw280wfqga5MfWQR8POLWzqYiqBfIqzYbu47HDcLcU' last in use 24655.608690999994 milliseconds ago
2021-08-04 09:30:19,474 Deleting 1 modules for <bokeh.document.document.Document object at 0x19f7b3130>
2021-08-04 09:30:19,511 Discarding session '11hBJN3s4Ikrlwa6PI6B9S7PWGKrGLWusGo9iqIjdy1D' last in use 24478.437876999997 milliseconds ago
2021-08-04 09:30:19,511 Deleting 1 modules for <bokeh.document.document.Document object at 0x19f7c5a90>
2021-08-04 09:30:19,544 Discarding session 'IuQbcH5WWZyB0t8KHr7Jb31XSOAIKQz29ibiJaxb9TID' last in use 24149.122876 milliseconds ago
2021-08-04 09:30:19,544 Deleting 1 modules for <bokeh.document.document.Document object at 0x19f7ed3d0>
2021-08-04 09:30:19,576 Discarding session '13bYddGD9ILYMTVJNLuj0gMRPFoRYZYsrIfT268wGBiY' last in use 24003.384987999998 milliseconds ago
2021-08-04 09:30:19,576 Deleting 1 modules for <bokeh.document.document.Document object at 0x19f807ca0>
2021-08-04 09:30:19,608 Discarding session '4mBbkvyR5XAZa0lYSdcGVSq2mQHuI1unhzXUHTpkrRJ8' last in use 23651.288306999995 milliseconds ago
2021-08-04 09:30:19,608 Deleting 1 modules for <bokeh.document.document.Document object at 0x19f79cbb0>
2021-08-04 09:30:19,641 Discarding session 'ekgBwUADBcrP2T2yLuMSVFL7C5lo0LaU0LpW9DCvwAxd' last in use 23495.205191 milliseconds ago
2021-08-04 09:30:19,641 Deleting 1 modules for <bokeh.document.document.Document object at 0x19f846490>
2021-08-04 09:30:19,674 Discarding session '2CbBGeA7Bp4vBbVC006MZT6pVUDPzS6YL1Zayb0e7h8P' last in use 23145.071225 milliseconds ago
2021-08-04 09:30:19,674 Deleting 1 modules for <bokeh.document.document.Document object at 0x19f85ee80>
2021-08-04 09:30:19,706 Discarding session '7zLiAGlnAAdgu5i6pZlGl8ACbV8zfUWnFDK1GvBckIo0' last in use 23008.255055 milliseconds ago
2021-08-04 09:30:19,706 Deleting 1 modules for <bokeh.document.document.Document object at 0x19f887790>
2021-08-04 09:30:19,739 Discarding session 'JnOu0FnBgn6omBckaj5WM1UCAm8HCuQtul9uAcgnIEQm' last in use 22660.284982 milliseconds ago
2021-08-04 09:30:19,739 Deleting 1 modules for <bokeh.document.document.Document object at 0x19f8ba070>
2021-08-04 09:30:19,774 Discarding session '4QzAKw05auxFmP4RjhbIkdIKEAHGt84sW0jgEYnrk9vA' last in use 22460.730046999997 milliseconds ago
2021-08-04 09:30:19,774 Deleting 1 modules for <bokeh.document.document.Document object at 0x19f8ca9d0>
2021-08-04 09:30:19,806 Discarding session '24oRPaxZCg4S8ZOW0im814Cq00QHiiYl5123MkF5tTR8' last in use 22289.793128 milliseconds ago
2021-08-04 09:30:19,806 Deleting 1 modules for <bokeh.document.document.Document object at 0x19f8f4370>
2021-08-04 09:30:19,838 Discarding session 'YDCzBFJXGMR0I8p5wZcB5jkatgVQdYFmMSTIviPQ93GX' last in use 20316.413202000003 milliseconds ago
2021-08-04 09:30:19,839 Deleting 1 modules for <bokeh.document.document.Document object at 0x19f910c70>
2021-08-04 09:30:21,232 [pid 99813] 0 clients connected
2021-08-04 09:30:21,232 [pid 99813]   /sliders has 0 sessions with 0 unused
2021-08-04 09:30:21,233 [pid 99813] Memory usage: 89.00 MB (RSS), 8261.00 MB (VMS)
2021-08-04 09:30:21,249   uncollected Documents: 2
2021-08-04 09:30:21,263   uncollected Sessions: 0
2021-08-04 09:30:21,283   uncollected Models: 1
2021-08-04 09:30:23,232 [pid 99813] 0 clients connected
2021-08-04 09:30:23,232 [pid 99813]   /sliders has 0 sessions with 0 unused
2021-08-04 09:30:23,233 [pid 99813] Memory usage: 90.00 MB (RSS), 8261.00 MB (VMS)
2021-08-04 09:30:23,249   uncollected Documents: 2
2021-08-04 09:30:23,263   uncollected Sessions: 0
2021-08-04 09:30:23,284   uncollected Models: 1
```
</details>

The server starts with  1 uncollected model (I forget what that is about but I think it's ok/expected) but after clearing all the closed session, does still report 2 uncollected documents. That's worth investigating but would not explain a failure to drop at all, since 15 other sessions/documents *were* collected. Even if there are cycles those should eventually be collected by gc. 

It's probably worth constructing an MRE that uses more memory, relative to the baseline process usage.

**Comment by EfremBraun** [comment]
>  It's probably worth constructing an MRE that uses more memory, relative to the baseline process usage.

The weather example (https://github.com/bokeh/bokeh/tree/branch-2.4/examples/app/weather) works well for this. The result I got from running the same protocol as above (with shorter times, since I wasn't feeling as patient...) is:
![Figure_1](https://user-images.githubusercontent.com/6618019/128223580-b6f32dab-a65b-4320-bcff-1b4f27c81daf.png)

**Comment by bryevdv** [comment]
One thing that wasn't a possibility in the past was adding `tracemalloc` logging of any kind, but now that we are Python 3.7+ we can use it from the standard library. I will see about augmenting the current memory logging with `tracemalloc`

**Comment by bryevdv** [comment]
@EfremBraun I have tried in several ways and cannot rally reconcile the results from `mprof`

### system `top`

I changed the example to create a 20M element array. Here is some output from `top` after a session is opened and then again after it is collected:

```
3632   python3.9    38.0  00:35.05 3      0    27    233M-  0B     0B     3631  3631  sleeping *0[1]              0.00000 0.00000    501  665012+    5499

3662   python3.9    52.3  01:17.81 3      0    27     67M+   0B     0B     3661  3661  sleeping *0[1]     
```
You can see the mem usage reported by `top` went down. from ~233MB to 67MB (and ~66MB is what was reported on cold startup)

### pympler

Another useful memory profiling tool is `pympler`. I had it print out the top 30 largest uses of memory with the example actually sending 200k element arrays to the client (~10 sessions). You can see before, during, and after session collection. The middle one clearly shows the `numpy` arrays in memory dominating as expected, then after collection they are gone:

<details>

```
[['str', 66378, 10958169],
 ['dict', 29685, 8985600],
 ['code', 18746, 3328065],
 ['type', 2856, 2596496],
 ['list', 22876, 1815856],
 ['tuple', 17470, 1073568],
 ['set', 480, 551168],
 ['bokeh.core.has_props.MetaHasProps', 329, 350184],
 ['weakref', 3944, 283968],
 ['_io.BufferedWriter', 2, 262480],
 ['int', 7856, 236432],
 ['wrapper_descriptor', 2899, 208728],
 ['getset_descriptor', 3161, 202304],
 ['abc.ABCMeta', 189, 195960],
 ['method_descriptor', 2691, 193752],
 ['builtin_function_or_method', 2630, 189360],
 ['frozenset', 404, 176864],
 ['function (__init__)', 1109, 150824],
 ['_io.BufferedReader', 1, 131240],
 ['cell', 3211, 128440],
 ['fused_cython_function', 483, 108192],
 ['property', 1224, 88128],
 ['jinja2.nodes.NodeType', 69, 74184],
 ['pytz.lazy.LazySet.__new__.<locals>.LazySet', 2, 65984],
 ['bokeh.core.property.descriptors.PropertyDescriptor', 1314, 63072],
 ['re.Pattern', 150, 62144],
 ['function (<lambda>)', 367, 49912],
 ['typing._TypedDictMeta', 45, 47880],
 ['cython_function_or_method', 235, 47000],
 ['_frozen_importlib.ModuleSpec', 966, 46368]]




[['numpy.ndarray', 57, 12806785],
 ['str', 67512, 11685163],
 ['dict', 33509, 9563288],
 ['code', 18906, 3366661],
 ['type', 2875, 2622560],
 ['list', 23530, 1871984],
 ['tuple', 18764, 1147176],
 ['set', 1719, 863592],
 ['bokeh.core.has_props.MetaHasProps', 334, 359240],
 ['weakref', 4136, 297792],
 ['_io.BufferedWriter', 2, 262480],
 ['int', 8041, 241728],
 ['wrapper_descriptor', 2899, 208728],
 ['getset_descriptor', 3187, 203968],
 ['builtin_function_or_method', 2771, 199512],
 ['abc.ABCMeta', 189, 197032],
 ['method_descriptor', 2691, 193752],
 ['frozenset', 404, 176864],
 ['bokeh.core.property.wrappers.PropertyValueDict', 668, 165664],
 ['function (__init__)', 1113, 151368],
 ['cell', 3470, 138800],
 ['collections.OrderedDict', 62, 133496],
 ['_io.BufferedReader', 1, 131240],
 ['fused_cython_function', 483, 108192],
 ['property', 1230, 88560],
 ['inspect.Parameter', 1357, 86848],
 ['jinja2.nodes.NodeType', 69, 74184],
 ['pytz.lazy.LazySet.__new__.<locals>.LazySet', 2, 65984],
 ['bokeh.core.property.descriptors.PropertyDescriptor', 1331, 63888],
 ['re.Pattern', 151, 62336]]




[['str', 67051, 11646236],
 ['dict', 30445, 9167464],
 ['code', 18903, 3366133],
 ['type', 2875, 2621408],
 ['list', 23345, 1858032],
 ['tuple', 17716, 1088552],
 ['set', 640, 630528],
 ['bokeh.core.has_props.MetaHasProps', 334, 359240],
 ['weakref', 4138, 297936],
 ['_io.BufferedWriter', 2, 262480],
 ['int', 7964, 239572],
 ['wrapper_descriptor', 2899, 208728],
 ['getset_descriptor', 3187, 203968],
 ['builtin_function_or_method', 2773, 199656],
 ['abc.ABCMeta', 189, 197032],
 ['method_descriptor', 2691, 193752],
 ['frozenset', 404, 176864],
 ['function (__init__)', 1113, 151368],
 ['cell', 3462, 138480],
 ['collections.OrderedDict', 62, 133496],
 ['_io.BufferedReader', 1, 131240],
 ['fused_cython_function', 483, 108192],
 ['property', 1230, 88560],
 ['inspect.Parameter', 1357, 86848],
 ['jinja2.nodes.NodeType', 69, 74184],
 ['pytz.lazy.LazySet.__new__.<locals>.LazySet', 2, 65984],
 ['bokeh.core.property.descriptors.PropertyDescriptor', 1331, 63888],
 ['re.Pattern', 151, 62336],
 ['function (<lambda>)', 373, 50728],
 ['typing._TypedDictMeta', 45, 47880]]
```
</details>

### stdlib `gc`

I also added addition printouts to show any uncollected Bokeh app modules, and uncollectable `gc.garbage`, and `gc.stats` as well.  All of these are also consistent with everything that is unreferenced being gone/collected. This shows before/after 7 closed sessions were collected:

```
2021-08-04 13:38:38,536 [pid 3771] 0 clients connected
2021-08-04 13:38:38,536 [pid 3771]   /sliders has 7 sessions with 7 unused
2021-08-04 13:38:38,536 [pid 3771] Memory usage: 1164.00 MB (RSS), 9037.00 MB (VMS)
2021-08-04 13:38:38,551   uncollected Documents: 8
2021-08-04 13:38:38,564   uncollected Sessions: 7
2021-08-04 13:38:38,586   uncollected Models: 253
2021-08-04 13:38:38,603   uncollected modules: 7
2021-08-04 13:38:38,603 [{'collections': 251, 'collected': 972, 'uncollectable': 0}, {'collections': 22, 'collected': 45, 'uncollectable': 0}, {'collections': 2, 'collected': 59, 'uncollectable': 0}]
2021-08-04 13:38:38,604 []
2021-08-04 13:38:38,604 Scheduling 7 sessions to discard
2021-08-04 13:38:38,604 Discarding session 'uE7vf9UQklPm3uHxqtw79GEks38VItI6RfHnGEhqHNFy' last in use 25825.539583 milliseconds ago
2021-08-04 13:38:38,604 Deleting 1 modules for <bokeh.document.document.Document object at 0x10b159a90>
2021-08-04 13:38:38,646 Discarding session 'CXE9nMpIvSPn3IXYbQXIAuopig5Y8n5g69kK711xqhFV' last in use 23025.724605000003 milliseconds ago
2021-08-04 13:38:38,646 Deleting 1 modules for <bokeh.document.document.Document object at 0x10b289fd0>
2021-08-04 13:38:38,688 Discarding session 'H12AOp0DQxZFPHxqgR7TZdH7lKuwhxFUvFKshE1TjYnn' last in use 22608.468289999997 milliseconds ago
2021-08-04 13:38:38,688 Deleting 1 modules for <bokeh.document.document.Document object at 0x10b2ccb80>
2021-08-04 13:38:38,732 Discarding session '0e5k9QTRKHmf6H87jMwguzcFqHGbSLKUNU28oPOUqfuv' last in use 21589.372654999996 milliseconds ago
2021-08-04 13:38:38,732 Deleting 1 modules for <bokeh.document.document.Document object at 0x10b2f3580>
2021-08-04 13:38:38,776 Discarding session 'aToh31DHWq31CPNWLh96WPLTc1SZrSqYMf8VJjAGM0qg' last in use 22231.570615000004 milliseconds ago
2021-08-04 13:38:38,776 Deleting 1 modules for <bokeh.document.document.Document object at 0x10b30de50>
2021-08-04 13:38:38,817 Discarding session 'kN8ERoEmQzbNPZyKxs0CaILiMzWO9fRksbnSNUFD1YwU' last in use 21975.774238000005 milliseconds ago
2021-08-04 13:38:38,817 Deleting 1 modules for <bokeh.document.document.Document object at 0x10b336370>
2021-08-04 13:38:38,857 Discarding session 'ImX9hkk6SjVIxXyVWmPFiIppeBDfTSdln1Fz3FELyDjr' last in use 20234.028847 milliseconds ago
2021-08-04 13:38:38,857 Deleting 1 modules for <bokeh.document.document.Document object at 0x10b3508e0>
2021-08-04 13:38:40,537 [pid 3771] 0 clients connected
2021-08-04 13:38:40,537 [pid 3771]   /sliders has 0 sessions with 0 unused
2021-08-04 13:38:40,537 [pid 3771] Memory usage: 1164.00 MB (RSS), 9037.00 MB (VMS)
2021-08-04 13:38:40,550   uncollected Documents: 2
2021-08-04 13:38:40,563   uncollected Sessions: 0
2021-08-04 13:38:40,583   uncollected Models: 1
2021-08-04 13:38:40,600   uncollected modules: 0
2021-08-04 13:38:40,600 [{'collections': 251, 'collected': 972, 'uncollectable': 0}, {'collections': 22, 'collected': 45, 'uncollectable': 0}, {'collections': 9, 'collected': 5157, 'uncollectable': 0}]
2021-08-04 13:38:42,599 []
```

The empty list is `gc.garbage` (nothing unreachable and uncollectable) The list of dicts is `gc.stats` which shows nothing uncollectable before or after, and shows after ~5k collection happened after sesssion cleanup. 


----

@EfremBraun can you elaborate what prompted you to run `mprof` in the first place? Can you point to any other evidence to support there being a memory leak? I will keep invesigating, but as it stands nothing seems to corroborate the `mprof` plot.

**Comment by EfremBraun** [comment]
I was originally my Mac's Activity Monitor (which is similar to Linux's `top`), which was showing me the only-increasing memory. I decided to use `mprof` based on its usage in #7468.

I'll try some of these other tools now and get back to you.

By the way, did you try `mprof` yourself? I'm thinking it's more likely to be some kind of Mac issue than an `mprof` issue. If you ran `mprof` and saw the same thing that I saw though, then that isn't the case.

**Comment by bryevdv** [comment]
I am seeing the same results as you in `mprof` I just can't get anything else to square with them.

**Comment by EfremBraun** [comment]
> @EfremBraun how exactly did you open new sessions in your brower?

By going to http://localhost:5006/slider

**Comment by EfremBraun** [comment]
> I am seeing the same results as you in `mprof` I just can't get anything else to square with them.

Ah, so it's not my Mac then...I thought I had seen some post on Discourse where a person was saying they were seeing a memory leak like this, but when they ran it on Heroku the issue disappeared, which is what led me to think the issue might be machine-specific. I can't find that post though.

**Comment by EfremBraun** [comment]
I tried `top`, and it's showing me the same result as `mprof`. Memory goes up and doesn't come back down when the sessions close. Not sure why I'm getting a different result from you on that one...

**Comment by bryevdv** [comment]
Well, here is what is weird. The RSS reported by `psutil` int he memory logging is not going down (agrees w/ mprof at least in trend) but the memory reported by `top` is absolutely going back down:

<img width="928" alt="Screen Shot 2021-08-04 at 14 14 07" src="https://user-images.githubusercontent.com/1078448/128255875-4851ff01-29c8-4f18-8520-b820d938cea0.png">

before dropping back to 70M the number in top was ~1GB after opening several sessions. I am honestly not sure which tools numbers to believe. 

> By going to http://localhost:5006/slider

I am asking for the literal exact physical process. Did you type or paste? Into one tab that you reloaded? Or in to several physical tabs?

**Comment by EfremBraun** [comment]
I just tried testing on my Ubuntu Virtual Box, and I see the same issue with `top` as on my Mac, which is that the memory goes up and doesn't come back down. Very strange how that's not what's going on with your system.

**Comment by bryevdv** [comment]
More experiments: I just printed `len(gc.get_objects())`

After opening a few sessions and letting them all close: `2021-08-04 14:25:20,695 103407`

After opening a dozen new sessions: `2021-08-04 14:25:32,709 111762` (i.e. went up quite a bit)

After letting all those sessions close: `2021-08-04 14:27:00,693 103449` (almost back to where it was)


I have to believe Python is telling the truth when it says it does not have those objects in its VM anymore. Is this relevant? 

https://distributed.dask.org/en/latest/worker.html#memory-not-released-back-to-the-os

> In many cases, high unmanaged memory usage or “memory leak” warnings on workers can be misleading: a worker may not actually be using its memory for anything, but simply hasn’t returned that unused memory back to the operating system, and is hoarding it just in case it needs the memory capacity again. This is not a bug in your code, nor in Dask — it’s actually normal behavior for all processes on Linux and MacOS, and is a consequence of how the low-level memory allocator works

**Comment by EfremBraun** [comment]
> In many cases, high unmanaged memory usage or “memory leak” warnings on workers can be misleading: a worker may not actually be using its memory for anything, but simply hasn’t returned that unused memory back to the operating system, and is hoarding it just in case it needs the memory capacity again. This is not a bug in your code, nor in Dask — it’s actually normal behavior for all processes on Linux and MacOS, and is a consequence of how the low-level memory allocator works

That might be the case. When I close all the tabs of the weather application, wait for Bokeh to delete the sessions, and then start re-opening sessions, the memory (as recorded by `top`) doesn't start to increase until I've opened up the same number of sessions before. So yes, perhaps Bokeh is just hoarding memory for re-use.

However, a different application of mine (the one I'm actually working on, rather than using it as a MWE), does not appear to be doing this. When I close the sessions and open new ones, the memory immediately begins going up as before. Let me look into this more...

**Comment by bryevdv** [comment]
In the mean time I will fire up a windows VM and see what shows up on a non-'nix platform

**Comment by EfremBraun** [comment]
Could you explain where in the code you were printing `len(gc.get_objects())`? I'm assuming it's somewhere in the Bokeh source code, but I can't figure out where it should go.

**Comment by bryevdv** [comment]
@EfremBraun I am adding various print outs to `_log_mem`

https://github.com/bokeh/bokeh/blob/9bfa131130da3a6905654b1f9f4388c87f1f91b0/bokeh/server/tornado.py#L708

Note you will need to set a non-zero ` --mem-log-frequency` and at least debut `--log-level`

**Comment by bryevdv** [comment]
On Windows the `psutil` RSS and VMS  memory numbers reported in the `_log_mem` do go  up and then 100% back down as sessions are cleared. There does still seem to be the 1 extra Model that sticks around same as on the other platforms.

**Comment by EfremBraun** [comment]
I'm seeing the same as what you're seeing (there only being 1 extra model not collected) for the weather application. This leads me to believe that you're indeed correct about the memory not being used, but just not being returned to the OS.

For my particular application, something else seems to be going on where many more models aren't being collected. Strangely, this doesn't happen if I don't attach the widgets to their callbacks. I know that's too ambiguous...I'll work on creating a MWE.

**Comment by EfremBraun** [comment]
I was able to isolate the issue, and it's quite puzzling. When a widget is connected to a callback with the `on_click` method, the garbage collection stops working.

Here's an example to illustrate. I took the weather example (https://github.com/bokeh/bokeh/tree/branch-2.4/examples/app/weather), and simply added in a button and connected that button to a callback function that immediately passed. This change results in the models present not getting collected. The full program (with a `#Change` comment on the few lines I added/changed) is here (remember to add the`2015_weather.csv` dataset to your directory before running):
```
import datetime
from os.path import dirname, join

import pandas as pd
from scipy.signal import savgol_filter

from bokeh.io import curdoc
from bokeh.layouts import column, row
from bokeh.models import ColumnDataSource, DataRange1d, Select, Button #Change: I added in an import of the Button widget
from bokeh.palettes import Blues4
from bokeh.plotting import figure

STATISTICS = ['record_min_temp', 'actual_min_temp', 'average_min_temp', 'average_max_temp', 'actual_max_temp', 'record_max_temp']

def get_dataset(src, name, distribution):
    df = src[src.airport == name].copy()
    del df['airport']
    df['date'] = pd.to_datetime(df.date)
    # timedelta here instead of pd.DateOffset to avoid pandas bug < 0.18 (Pandas issue #11925)
    df['left'] = df.date - datetime.timedelta(days=0.5)
    df['right'] = df.date + datetime.timedelta(days=0.5)
    df = df.set_index(['date'])
    df.sort_index(inplace=True)
    if distribution == 'Smoothed':
        window, order = 51, 3
        for key in STATISTICS:
            df[key] = savgol_filter(df[key], window, order)

    return ColumnDataSource(data=df)

def make_plot(source, title):
    plot = figure(x_axis_type="datetime", width=800, tools="", toolbar_location=None)
    plot.title.text = title

    plot.quad(top='record_max_temp', bottom='record_min_temp', left='left', right='right',
              color=Blues4[2], source=source, legend_label="Record")
    plot.quad(top='average_max_temp', bottom='average_min_temp', left='left', right='right',
              color=Blues4[1], source=source, legend_label="Average")
    plot.quad(top='actual_max_temp', bottom='actual_min_temp', left='left', right='right',
              color=Blues4[0], alpha=0.5, line_color="black", source=source, legend_label="Actual")

    # fixed attributes
    plot.xaxis.axis_label = None
    plot.yaxis.axis_label = "Temperature (F)"
    plot.axis.axis_label_text_font_style = "bold"
    plot.x_range = DataRange1d(range_padding=0.0)
    plot.grid.grid_line_alpha = 0.3

    return plot

def update_plot(attrname, old, new):
    city = city_select.value
    plot.title.text = "Weather data for " + cities[city]['title']

    src = get_dataset(df, cities[city]['airport'], distribution_select.value)
    source.data.update(src.data)

def test_function(): #Change: Added this function
    pass

city = 'Austin'
distribution = 'Discrete'

cities = {
    'Austin': {
        'airport': 'AUS',
        'title': 'Austin, TX',
    },
    'Boston': {
        'airport': 'BOS',
        'title': 'Boston, MA',
    },
    'Seattle': {
        'airport': 'SEA',
        'title': 'Seattle, WA',
    }
}

city_select = Select(value=city, title='City', options=sorted(cities.keys()))
distribution_select = Select(value=distribution, title='Distribution', options=['Discrete', 'Smoothed'])
test_button = Button(label='test') #Change: Added in this button

df = pd.read_csv(join(dirname(__file__), 'data/2015_weather.csv'))
source = get_dataset(df, cities[city]['airport'], distribution)
plot = make_plot(source, "Weather data for " + cities[city]['title'])

city_select.on_change('value', update_plot)
distribution_select.on_change('value', update_plot)
test_button.on_click(test_function) #Change: Connected the button to the function

controls = column(city_select, distribution_select, test_button) #Change: Added the button to the layout

curdoc().add_root(row(plot, controls))
curdoc().title = "Weather"
```

I see 53 uncollected models after starting a session and then ending it. If I then start another new session and end it though, the number of uncollected models remains at 53.

So this appears to be showing some strange kind of memory leak.

**Comment by bryevdv** [comment]
@EfremBraun Thanks for digging into things and making the connection. I know the there are some places in the dask dashboard where apparently they jump through some hoops to prevent memory cycles specifically related to callbacks, so I will start looking in this area.

**Comment by bryevdv** [comment]
I've made some inroads in this, I think I can see were all the callbacks are introducing cyclic dependencies. Some quick experiments with adding `weakref` seem promising but will probably need iteration. I also think the one "extra" document is getting held by the global "current state" which holds "curdoc". That will also require some thought and cleanup since there really should not be a *global* "curdoc" in the case of a Bokeh server ("curdoc" should be per session)

**Comment by bryevdv** [comment]
Just also noting this for posterity as well: https://github.com/holoviz/panel/issues/2640#issuecomment-909341786

TLDR importing numba can cause a leak, and produce the "extra referrers" error message. This will have to be fixed on the numba side of things.
